From 2be9e12f81fecb4a0f5b34d0284919b4e267cce4 Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <thomas.langer@intel.com>
Date: Fri, 17 Jul 2020 19:53:50 +0200
Subject: [PATCH] PONRTSYS-7614: xpcs: Fix apply same speed for XAUI mode

Also accept the same speed for the ONEG_XAUI mode.

Fixes: 0db1ad012c5e (PONRTSYS-7614: xpcs: Allow to apply current
settings)
---
 drivers/net/ethernet/lantiq/xpcs/xpcs.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.c b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
index b17f77c2807e..fdfada0ad2a6 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.c
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
@@ -1563,7 +1563,7 @@ static int xpcs_ethtool_ksettings_set(struct device *dev,
 	}
 
 	if ((speed == SPEED_1000 || speed == SPEED_100 || speed == SPEED_10) &&
-	    (pdata->conntype == CONN_TYPE_ANEG_DIS && pdata->speed != speed))
+	     pdata->conntype == CONN_TYPE_ANEG_DIS)
 		mode = ONEG_XAUI_MODE;
 	else if (speed == SPEED_10000)
 		mode = TENG_KR_MODE;
@@ -1572,11 +1572,11 @@ static int xpcs_ethtool_ksettings_set(struct device *dev,
 	else
 		return -EINVAL;
 
-	pdata->speed = speed;
-
-	/* Restart Xpcs & PHY */
-	if (mode != pdata->mode)
+	/* Restart Xpcs & PHY if mode or speed changes */
+	if (mode != pdata->mode || speed != pdata->speed) {
+		pdata->speed = speed;
 		xpcs_reinit(pdata->dev, mode);
+	}
 
 	return 0;
 }
